torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -0,0 +1,295 @@
+  }
+  workCV_.notify_all();
+}
+
+const std::exception& ProcessGroupMPI::WorkMPI::exception() const {
+  return *ex_;
+}
+
+// ProcessGroupMPI
+
+// Static global states
+int ProcessGroupMPI::numProcessGroups_ = 0;
+int ProcessGroupMPI::mpiThreadSupport_ = 0;
+bool ProcessGroupMPI::mpiInitialized_ = false;
+std::mutex ProcessGroupMPI::pgGlobalMutex_;
+
+ProcessGroupMPI::ProcessGroupMPI()
+: ProcessGroup(-1, 0)
+, stop_(false)
+{
+  initMPI();
+  workerThread_ = std::thread(&ProcessGroupMPI::runLoop, this);
+}
+
+ProcessGroupMPI::~ProcessGroupMPI() {
+  destroy();
+}
+
+void ProcessGroupMPI::initMPI() {
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  // We only want to initialize once
+  try {
+    if (!mpiInitialized_) {
+      MPI_CHECK(MPI_Init_thread(nullptr,
+                                nullptr,
+                                MPI_THREAD_MULTIPLE,
+                                &mpiThreadSupport_));
+      if (mpiThreadSupport_ < MPI_THREAD_SERIALIZED) {
+        throw std::runtime_error("Used MPI implementation doesn't have the "
+                                 "minimum level of threading support: "
+                                 "MPI_THREAD_SERIALIZED. This is required by "
+                                 "c10d package");
+      }
+      if (mpiThreadSupport_ != MPI_THREAD_MULTIPLE) {
+        std::cerr << "WARNING: Used MPI implementation doesn't support "
+                  << "multithreading, so distributed cannot support more "
+                  << "than one process group work."
+                  << std::endl;
+      }
+      mpiInitialized_ = true;
+    }
+
+    // Update the world size and rank
+    MPI_CHECK(MPI_Comm_size(MPI_COMM_WORLD, &size_));
+    MPI_CHECK(MPI_Comm_rank(MPI_COMM_WORLD, &rank_));
+
+  } catch (...) {
+    MPI_CHECK(MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE));
+  }
+
+  if (mpiThreadSupport_ != MPI_THREAD_MULTIPLE && numProcessGroups_ >= 1) {
+    throw std::runtime_error("More than one process group created, "
+                             "this is not supported due to the used MPI "
+                             "implementation doesn't provide the full support "
+                             "of multi-threading");
+  }
+  // increase the total PG count
+  ++numProcessGroups_;
+}
+
+void ProcessGroupMPI::destroy() {
+  std::unique_lock<std::mutex> lock(pgMutex_);
+
+  while (!queue_.empty()) {
+    queueConsumeCV_.wait(lock);
+  }
+  // Queue is empty, signal stop
+  stop_ = true;
+
+  // Release lock to allow threads to terminate
+  queueProduceCV_.notify_all();
+
+  lock.unlock();
+
+  // Join the single worker thread
+  workerThread_.join();
+
+  // Decrease the number of PG created
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  --numProcessGroups_;
+}
+
+void ProcessGroupMPI::abort() {
+  destroy();
+  MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
+}
+
+void ProcessGroupMPI::finalize() {
+  MPI_CHECK(MPI_Finalize());
```

Comment:
Why can't this be called from `destroy()`, once the number of process groups reaches zero? The fact that you have to call finalize breaks the process group contract, since nothing else requires that.

Reply:
Alternatively, register an atexit handler